README.md
# here_transit

Asynchronous Python client for the HERE Transit V8 API

[![GitHub Actions](https://github.com/eifinger/here_transit/workflows/CI/badge.svg)](https://github.com/eifinger/here_transit/actions?workflow=CI)
[![PyPi](https://img.shields.io/pypi/v/here_transit.svg)](https://pypi.python.org/pypi/here_transit)
[![PyPi](https://img.shields.io/pypi/l/here_transit.svg)](https://github.com/eifinger/here_transit/blob/master/LICENSE)
[![codecov](https://codecov.io/gh/eifinger/here_transit/branch/master/graph/badge.svg)](https://codecov.io/gh/eifinger/here_transit)
[![Downloads](https://pepy.tech/badge/here_transit)](https://pepy.tech/project/here_transit)

## Installation

```bash
pip install here_transit
```

## Usage

```python
import asyncio

from here_transit import HERETransitApi, Place, Return

API_KEY = "<YOUR_API_KEY>"


async def main() -> None:
    """Show example how to get location of your tracker."""
    async with HERETransitApi(api_key=API_KEY) as here_transit:
        response = await here_transit.route(
            origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
            destination=Place(latitude=50.060940891421765, longitude=8.336477279663088),
            return_values=[Return.TRAVEL_SUMMARY],
        )
        print(
            f"Duration is: {response['routes'][0]['sections'][0]['travelSummary']['duration']}"
        )


if __name__ == "__main__":
    asyncio.run(main())
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.2.0 (2022-12-12)
### Feature
* Add HERETransitTooManyRequestsError ([`4d59ffb`](https://github.com/eifinger/here_transit/commit/4d59ffba449209d6569cfd1c6f3a4eccd5dc6ba2))

## v1.1.1 (2022-12-09)
### Fix
* Publish Exceptions as top level ([`f37377b`](https://github.com/eifinger/here_transit/commit/f37377b9c7298d64dd0736c87c8025393edd23e6))

## v1.1.0 (2022-12-09)
### Feature
* Add dedicated departureArrivalTooClose exception ([`86c631d`](https://github.com/eifinger/here_transit/commit/86c631d458e5af6ed7e0373e54fd7a0901d1d2b6))
* Add dedicated noRouteFound exception ([`7aa0fc3`](https://github.com/eifinger/here_transit/commit/7aa0fc314d41a8198f1611ec84de4a2e6be8f71c))
* Add dedicated NoTransitRouteFound exception ([`40d3bdc`](https://github.com/eifinger/here_transit/commit/40d3bdc89adbeebed5c2e730241cebd669f6062e))

## v1.0.0 (2022-10-21)
### Breaking
* Remove Python3.8 as supported version ([`a5966b6`](https://github.com/eifinger/here_transit/commit/a5966b6451166b47d75740c111bf2b835568c67d))

## v0.1.0 (2022-07-21)
### Feature
* Initial commit ([`845d596`](https://github.com/eifinger/here_transit/commit/845d5962e400443efe4918fe1361937f756ee614))

### Documentation
* Remove mentions of Routing API ([`8d47240`](https://github.com/eifinger/here_transit/commit/8d4724071203c9dc04128ffb2ea6e6a152b7f123))


tests/__init__.py
"""Tests for Asynchronous Python client for the HERE Transit V8 API."""


tests/test_here_transit.py
"""Tests for `here_transit.here_transit`."""
import os
from datetime import datetime

import aiohttp
import pytest

from here_transit.exceptions import (
    HERETransitDepartureArrivalTooCloseError,
    HERETransitError,
    HERETransitNoRouteFoundError,
    HERETransitNoTransitRouteFoundError,
    HERETransitTooManyRequestsError,
    HERETransitUnauthorizedError,
)
from here_transit.here_transit import API_HOST, API_VERSION, ROUTES_PATH, HERETransitApi
from here_transit.model import Place, Return, TransitMode


@pytest.mark.asyncio
async def test_car_route(aresponses):
    """Test getting a route."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("transit_route_response.json"),
            status=200,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        here_api = HERETransitApi(api_key="key", session=session)
        response = await here_api.route(
            included_modes=[TransitMode.BUS],
            origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
            destination=Place(latitude=50.060940891421765, longitude=8.336477279663088),
            return_values=[Return.TRAVEL_SUMMARY, Return.POLYLINE],
            departure_time=datetime.now(),
        )

        assert response["routes"][0]["sections"][0]["travelSummary"]["duration"] == 1140


@pytest.mark.asyncio
async def test_invalid_key(aresponses):
    """Test that an invalid api_key throws HERETransitUnauthorizedError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("invalid_key_response.json"),
            status=401,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERETransitUnauthorizedError):
            here_api = HERETransitApi(api_key="invalid", session=session)
            await here_api.route(
                origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )


@pytest.mark.asyncio
async def test_invalid_request(aresponses):
    """Test that a invalid request throws HERETransitError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("invalid_request_response.json"),
            status=400,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERETransitError) as error:
            here_api = HERETransitApi(api_key="key", session=session)
            await here_api.route(
                origin=Place(latitude=150.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )
        assert "'origin': value is out of range" in str(error.value)


@pytest.mark.asyncio
async def test_429_too_many_requests(aresponses):
    """Test that a invalid request throws HERETransitTooManyRequestsError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("too_many_requests_response.json"),
            status=429,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERETransitTooManyRequestsError) as error:
            here_api = HERETransitApi(api_key="key", session=session)
            await here_api.route(
                origin=Place(latitude=150.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )
        assert "Rate limit for this service has been reached" in str(error.value)


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "json_fixture,expected_exception,expected_exception_message",
    [
        (
            "noRouteFound_response.json",
            HERETransitNoRouteFoundError,
            (
                "Route between origin and destination "
                "is not possible given current input parameters"
            ),
        ),
        (
            "noTransitRouteFound_response.json",
            HERETransitNoTransitRouteFoundError,
            "Transit routing between origin and destination is not possible",
        ),
        (
            "departureArrivalTooClose_response.json",
            HERETransitDepartureArrivalTooCloseError,
            "Departure and arrival locations are too close",
        ),
    ],
)
async def test_exceptions(
    aresponses, json_fixture, expected_exception, expected_exception_message
):
    """Test correct HERETransitError are thrown."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture(json_fixture),
            status=200,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(expected_exception) as error:
            here_api = HERETransitApi(api_key="key", session=session)
            await here_api.route(
                origin=Place(latitude=150.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
                changes=0,
            )
        assert expected_exception_message in str(error.value)


def load_json_fixture(filename: str) -> str:
    """Load a fixture."""
    path = os.path.join(os.path.dirname(__file__), "fixtures", filename)
    with open(path, encoding="utf-8") as fptr:
        return fptr.read()


here_transit/model.py
"""HERE Transit API models."""

from dataclasses import dataclass
from enum import Enum


class TransitMode(Enum):
    """Available TransportModes."""

    HIGHSPEED_TRAIN = "highSpeedTrain"
    INTERCITY_TRAIN = "intercityTrain"
    INTERREGIONAL_TRAIN = "interRegionalTrain"
    REGIONAL_TRAIN = "regionalTrain"
    CITY_TRAIN = "cityTrain"
    BUS = "bus"
    FERRY = "ferry"
    SUBWAY = "subway"
    LIGHT_RAIL = "lightRail"
    PRIVATE_BUS = "privateBus"
    INCLINED = "inclined"
    AERIAL = "aerial"
    BUS_RAPID = "busRapid"
    MONORAIL = "monorail"
    FLIGHT = "flight"
    SPACESHIP = "spaceship"


class UnitSystem(Enum):
    """Available UnitSystem Values."""

    METRIC = "metric"
    IMPERIAL = "imperial"


class Return(Enum):
    """Available Return Values."""

    INTERMEDIATE = "intermediate"
    FARES = "fares"
    POLYLINE = "polyline"
    ACTIONS = "actions"
    TRAVEL_SUMMARY = "travelSummary"
    INCIDENTS = "incidents"
    BOOKING_LINKS = "bookingLinks"


@dataclass
class Place:
    """Place for route requests."""

    latitude: float
    longitude: float


here_transit/here_transit.py
"""A module to query the HERE Transit API v8."""
from __future__ import annotations

import asyncio
import json
import socket
from datetime import datetime
from importlib import metadata
from typing import Any, Dict, List, MutableMapping

import aiohttp
import async_timeout
from yarl import URL

from .exceptions import (
    HERETransitConnectionError,
    HERETransitDepartureArrivalTooCloseError,
    HERETransitError,
    HERETransitNoRouteFoundError,
    HERETransitNoTransitRouteFoundError,
    HERETransitTooManyRequestsError,
    HERETransitUnauthorizedError,
)
from .model import Place, Return, TransitMode, UnitSystem

SCHEME = "https"
API_HOST = "transit.router.hereapi.com"
API_VERSION = "/v8"
ROUTES_PATH = "routes"


class HERETransitApi:
    """Main class for handling connections with the HERE Transit API v8."""

    def __init__(
        self,
        api_key: str,
        request_timeout: int = 10,
        session: aiohttp.client.ClientSession | None = None,
        user_agent: str | None = None,
    ) -> None:
        """Initialize connection with HERE Transit.
        Class constructor for setting up an HERETransitApi object to
        communicate with the HERE Transit API.
        Args:
            api_key: A key generated specifically to authenticate API requests.
            request_timeout: Max timeout to wait for a response from the API.
            session: Optional, shared, aiohttp client session.
            user_agent: Defaults to here_transit/<version>.
        """
        self._session = session
        self._api_key = api_key
        self._close_session = False

        self.request_timeout = request_timeout
        self.user_agent = user_agent

        if user_agent is None:
            version = metadata.version(__package__)
            self.user_agent = f"here_transit/{version}"

    async def request(
        self,
        uri: str,
        params: MutableMapping[str, str | List[str]],
        method: str = "GET",
    ) -> Any:
        """Handle a request to the HERE Transit API.
        Make a request against the HERE Transit API and handles the response.
        Args:
            uri: The request URI on the HERE Transit API to call.
            method: HTTP method to use for the request; e.g., GET, POST.
            data: RAW HTTP request data to send with the request.
            json_data: Dictionary of data to send as JSON with the request.
            params: Mapping of request parameters to send with the request.
        Returns:
            The response from the API. In case the response is a JSON response,
            the method will return a decoded JSON response as a Python
            dictionary. In other cases, it will return the RAW text response.
        Raises:
            HERETransitConnectionError: An error occurred while communicating
                with the HERE Transit API (connection issues).
            HERETransitError: An error occurred while processing the
                response from the HERE Transit API (invalid data).
        """
        url = URL.build(scheme=SCHEME, host=API_HOST, path=API_VERSION) / uri

        headers = {
            "User-Agent": self.user_agent,
            "Accept": "application/json",
            "DNT": "1",
        }
        params["apiKey"] = self._api_key

        if self._session is None:
            self._session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self._session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                )
        except asyncio.TimeoutError as exception:
            raise HERETransitConnectionError(
                "Timeout occurred while connecting to the HERE Transit API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise HERETransitConnectionError(
                "Error occurred while communicating with the HERE Transit API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        contents = await response.read()
        decoded_contents = contents.decode("utf8")
        if "application/json" not in content_type:
            raise HERETransitError(response.status, {"message": decoded_contents})
        if response.status // 100 in [4, 5]:
            response.close()

            if response.status == 401:
                raise HERETransitUnauthorizedError(
                    json.loads(decoded_contents)["error_description"]
                )
            if response.status == 429:
                raise HERETransitTooManyRequestsError(
                    json.loads(decoded_contents)["error_description"]
                )

            raise HERETransitError(response.status, json.loads(decoded_contents))

        return await response.json()

    async def route(  # pylint: disable=too-many-arguments,too-many-locals
        self,
        origin: Place,
        destination: Place,
        pedestrian_peed: int = 1,
        pedestrian_max_distance: int = 2000,
        alternatives: int = 0,
        units: UnitSystem = UnitSystem.METRIC,
        lang: str = "en-US",
        changes: int | None = None,
        included_modes: List[TransitMode] | None = None,
        excluded_modes: List[TransitMode] | None = None,
        return_values: List[Return] | None = None,
        departure_time: datetime | None = None,
        arrival_time: datetime | None = None,
    ) -> Any:
        """Get the route.
        Args:
            origin: Latitude and longitude of the origin.
            destination: Latitude and longitude of the destination.
            pedestrianSpeed: Walking speed in meters per second.
            pedestrianMaxDistance: Maximum allowed walking distance in meters.
            alternatives: Number of alternative routes to return.
            units: Unitsystem to use.
            lang: IETF BCP 47 compatible language identifier.
            changes: Maximum number of changes/transfers. None means unlimited.
            included_modes: Included TransitModes. None means all.
            excluded_modes: Excluded TransitModes. None means none.
            return_values: HERE Transit API return values to request.
            departure_time: Departure time.
            arrival_time: Arrival time.
        Returns:
            The response from the API.
        Raises:
            HERETransitConnectionError: An error occurred while communicating
                with the HERE Transit API (connection issues).
            HERETransitError: An error occurred while processing the
                response from the HERE Transit API (invalid data).
        """
        if included_modes is not None and excluded_modes is not None:
            if len(included_modes) > 0 and len(excluded_modes) > 0:
                raise ValueError("Cannot include and exclude at the same time")

        params: MutableMapping[str, str | List[str]] = {  # type: ignore
            "origin": f"{origin.latitude},{origin.longitude}",
            "destination": f"{destination.latitude},{destination.longitude}",
            "pedestrianSpeed": str(pedestrian_peed),
            "pedestrianMaxDistance": str(pedestrian_max_distance),
            "alternatives": str(alternatives),
            "units": units.value,
            "lang": lang,
        }
        if changes is not None:
            params["changes"] = str(changes)
        if included_modes is not None:
            params["modes"] = ",".join(m.value for m in included_modes)
        if excluded_modes is not None:
            params["modes"] = ",".join(f"-{m.value}" for m in excluded_modes)
        if return_values is not None:
            params["return"] = ",".join(r.value for r in return_values)
        if departure_time is not None:
            params["departureTime"] = departure_time.isoformat(timespec="seconds")
        if arrival_time is not None:
            params["arrivalTime"] = arrival_time.isoformat(timespec="seconds")

        response = await self.request(uri=ROUTES_PATH, params=params)

        if len(response["routes"]) < 1:
            raise_error_from_notices(response["notices"])
        return response

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> HERETransitApi:
        """Async enter.
        Returns:
            The HERETransitApi object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.
        Args:
            _exc_info: Exec type.
        """
        await self.close()


def raise_error_from_notices(notices: List[Dict[str, str]]) -> None:
    """Raise the correct error for the contained notices."""
    errors = {notice["code"]: notice["title"] for notice in notices}

    if (title := errors.get("departureArrivalTooClose")) is not None:
        raise HERETransitDepartureArrivalTooCloseError(title)
    if (title := errors.get("noTransitRouteFound")) is not None:
        raise HERETransitNoTransitRouteFoundError(title)
    if (title := errors.get("noRouteFound")) is not None:
        raise HERETransitNoRouteFoundError(title)

    raise HERETransitError(",".join(errors.values()))
